src/serac/coefficients/sensitivity_coefficients.cpp

```diff
@@ -0,0 +1,110 @@
+#include "serac/physics/utilities/physics_utils.hpp"
+
+namespace serac::mfem_ext {
+
+ShearSensitivityCoefficient::ShearSensitivityCoefficient(FiniteElementState&    displacement,
+                                                         FiniteElementState&    adjoint_displacement,
+                                                         LinearElasticMaterial& linear_mat)
+    : displacement_(displacement), adjoint_displacement_(adjoint_displacement), material_(linear_mat)
+{
+  dim_ = displacement_.gridFunc().ParFESpace()->GetVDim();
+
+  du_dX_.SetSize(dim_);
+  dp_dX_.SetSize(dim_);
+  adjoint_strain_.SetSize(dim_);
+  d_sigma_d_shear_.SetSize(dim_);
+  F_.SetSize(dim_);
+}
+
+double ShearSensitivityCoefficient::Eval(mfem::ElementTransformation& T, const mfem::IntegrationPoint& ip)
+{
+  T.SetIntPoint(&ip);
+
+  // Compute the displacement and deformation gradient
+  displacement_.gridFunc().GetVectorGradient(T, du_dX_);
+
+  solid_util::calcDeformationGradient(du_dX_, F_);
+
+  double det_J = F_.Det();
+
+  // Compute the adjoint gradient and strain
+  adjoint_displacement_.gridFunc().GetVectorGradient(T, dp_dX_);
+
+  serac::solid_util::calcLinearizedStrain(dp_dX_, adjoint_strain_);
+
+  // Evaluate the derivative of the Cauchy stress with respect to the shear modulus using the calculated deformation
+  // gradient
+  material_.EvalShearSensitivity(du_dX_, d_sigma_d_shear_);
+
+  // Scale the derivative matrix by the geometric and quadrature weights
+  d_sigma_d_shear_ *= det_J * ip.weight * T.Weight();
+
+  double scalar_sum = 0.0;
+
+  // Compute the inner product of the stress sensitivity and the adjoint strain
+  for (int i = 0; i < dim_; ++i) {
+    for (int j = 0; j < dim_; ++j) {
+      scalar_sum += d_sigma_d_shear_(i, j) * adjoint_strain_(i, j);
+    }
+  }
+
+  return scalar_sum;
+}
+
+BulkSensitivityCoefficient::BulkSensitivityCoefficient(FiniteElementState&    displacement,
+                                                       FiniteElementState&    adjoint_displacement,
+                                                       LinearElasticMaterial& linear_mat)
+    : displacement_(displacement), adjoint_displacement_(adjoint_displacement), material_(linear_mat)
+{
+  dim_ = displacement_.gridFunc().ParFESpace()->GetVDim();
+
+  du_dX_.SetSize(dim_);
+  dp_dX_.SetSize(dim_);
+  adjoint_strain_.SetSize(dim_);
+  d_sigma_d_bulk_.SetSize(dim_);
+  F_.SetSize(dim_);
+}
+
+double BulkSensitivityCoefficient::Eval(mfem::ElementTransformation& T, const mfem::IntegrationPoint& ip)
+{
+  T.SetIntPoint(&ip);
+
+  // Compute the displacement and deformation gradient
+  displacement_.gridFunc().GetVectorGradient(T, du_dX_);
+
+  solid_util::calcDeformationGradient(du_dX_, F_);
+
+  double det_J = F_.Det();
+
+  // Compute the adjoint gradient and strain
+  adjoint_displacement_.gridFunc().GetVectorGradient(T, dp_dX_);
+
+  serac::solid_util::calcLinearizedStrain(dp_dX_, adjoint_strain_);
```

Comment:
It seems like `adjoint_strain_`, `d_sigma_d_bulk_`, and `d_sigma_d_shear_` are temporaries based on the integration point and also `private` to the class.  Perhaps the code could be a little simpler if we just declared a temporary `DenseMatrix` here directly instead of the the `private` member.

Reply:
Also similar considerations for `du_dX`, `dp_dx`, and `F`.

Reply:
The issue is that it's fairly expensive to construct and destruct these dense matricies at each quadrature point as it uses dynamic allocations. This mutable private member solution is what is used in most of the MFEM integrators for that reason.